text/0000-router-service.md

```diff
@@ -148,11 +148,48 @@ Example: this feature can replace [this use of private API in ember-href-to](htt
 
 Takes a string URL and returns a promise that resolves to a `RouteInfoWithAttributes` for the leafmost route represented by the URL. The promise rejects if the URL is not recognized or an unhandled exception is encountered. This method expects to receive the actual URL as seen by the browser _including_ the app's `rootURL`.
 
-### Improved Event Coverage
+### Deprecating willTransition and didTransition
 
 Application-wide transition monitoring events belong on the Router service, not spread throughout the Route classes. That is the reason for the existing `willTransition` and `didTransition` hooks/events on the Router. But they are not sufficient to capture all the detail people need. See for example, https://github.com/nickiaconis/rfcs/blob/1bd98ec534441a38f62a48599ffa8a63551b785f/text/0000-transition-hooks-events.md
 
-I do not believe an additional event is needed, we just need to ensure that the existing two events fire at every appropriate time. `willTransition` should fire _whenever_ the destination route changes, and that includes redirects, error handling, and loading states. A long chain of redirects should cause a sequence of `willTransition` events and then a final `didTransition` event.
+In addition, they receive handlerInfos in their arguments, which are an undocumented internal implementation detail of router.js that doesn't belong in Ember's public API. Everything you can do with handlerInfos can be done with the RouteInfo type that is proposed in this RFC, with the benefit of sticking to supported public API.
+
+So we should deprecate willTransition and didTransition in favor of the following new events.
+
+### New Events: routeWillChange & routeDidChange
+
+The `routeWillChange` event fires whenever a new route is chosen as the desired target of a transition. This includes `transitionTo`, `replaceWith`, all redirection for any reason including error handling, and abort. Aborting implies changing the desired target back to where you already were. Once a transition has completed, `routeDidChange` fires. 
+
+Both events receive a single `transition` argument as described in the "Transition Object" section below, which explains the meaning of `from` and `to` in more detail.
+
+Redirection example:
+
+ 1. current route is A
+ 2. user initiates a transition to B
+ 3. routeWillChange fires `from` A `to` B.
+ 4. B redirects to C
+ 5. routeWillChange fires `from` A `to` C.
+ 6. routeDidChange fires `from` A `to` C.
+
+Abort example:
+
+ 1. current route is A
+ 2. user initiates a transition to B
+ 3. routeWillChange fires `from` A `to` B.
+ 4. in response to the previous routeWillChange event, the transition is aborted.
+ 5. routeWillChange fires `from` A `to` A.
+ 8. routeDidChange fires `from` A `to` A.
```

Comment:
The stream of events for each is distinct.

    willTransition A->B
    willTransition A->A
    didTransition A->A

vs

    willTransition A->A
    didTransition A->A

You can uniquely identify these sequences because they always end with didTransition.